File: analytics/drivers.java
import org.graalvm.polyglot.Context;
import org.graalvm.polyglot.Value;

import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class DriverCalculator {

    public static Object computeDrivers(String formula,
                                        Map<String, Map<Integer, Map<Integer, Double>>> data,
                                        boolean returnPercent) {

        // 1. Разбор формулы
        String[] parts = formula.split("=", 2);
        String leftStr = parts[0].trim();
        String rightStr = parts[1].trim();

        // 2. Маппинг имён -> алиасы
        Map<String, String> nameToAlias = new LinkedHashMap<>();
        Map<String, String> aliasToName = new LinkedHashMap<>();
        int idx = 0;
        for (String name : data.keySet()) {
            String alias = "VAR_" + idx++;
            nameToAlias.put(name, alias);
            aliasToName.put(alias, name);
        }
        if (!nameToAlias.containsKey(leftStr)) {
            nameToAlias.put(leftStr, "TARGET_VAR");
            aliasToName.put("TARGET_VAR", leftStr);
        }

        // 3. Замена имён в формуле (от длинных к коротким)
        List<String> sortedNames = new ArrayList<>(nameToAlias.keySet());
        sortedNames.sort((a, b) -> Integer.compare(b.length(), a.length()));
        String patternStr = String.join("|", sortedNames.stream()
                .map(Pattern::quote)
                .toArray(String[]::new));
        Pattern pattern = Pattern.compile(patternStr);
        Matcher matcher = pattern.matcher(rightStr);
        StringBuilder sb = new StringBuilder();
        while (matcher.find()) {
            matcher.appendReplacement(sb, nameToAlias.get(matcher.group()));
        }
        matcher.appendTail(sb);
        String safeRight = sb.toString();

        // Модификация оператора % в цепочку JS
        safeRight = safeRight.replace(" % ", " / 100 * ");

        // 4. Переводим данные в алиасы
        Map<String, Map<Integer, Map<Integer, Double>>> safeData = new LinkedHashMap<>();
        for (Map.Entry<String, Map<Integer, Map<Integer, Double>>> entry : data.entrySet()) {
            safeData.put(nameToAlias.get(entry.getKey()), entry.getValue());
        }

        // 5. Определяем драйверы
        Set<String> driversSet = new LinkedHashSet<>();
        for (String alias : nameToAlias.values()) {
            if (!alias.equals("TARGET_VAR") && safeData.containsKey(alias)) {
                driversSet.add(alias);
            }
        }
        List<String> drivers = new ArrayList<>(driversSet);
        if (drivers.isEmpty()) {
            throw new IllegalArgumentException("В формуле нет переменных.");
        }

        // 6. Временная шкала (пересечение периодов)
        Set<String> timeStrings = null;
        for (String v : drivers) {
            Set<String> vTimes = new HashSet<>();
            Map<Integer, Map<Integer, Double>> vData = safeData.get(v);
            if (vData != null) {
                for (int year : vData.keySet()) {
                    for (int month : vData.get(year).keySet()) {
                        vTimes.add(year + "-" + month);
                    }
                }
            }
            if (timeStrings == null) {
                timeStrings = vTimes;
            } else {
                timeStrings.retainAll(vTimes);
            }
        }

        List<int[]> timeKeys = new ArrayList<>();
        if (timeStrings != null) {
            List<String> sortedTimes = new ArrayList<>(timeStrings);
            sortedTimes.sort((a, b) -> {
                String[] p1 = a.split("-");
                String[] p2 = b.split("-");
                int yComp = Integer.compare(Integer.parseInt(p1[0]), Integer.parseInt(p2[0]));
                if (yComp != 0) return yComp;
                return Integer.compare(Integer.parseInt(p1[1]), Integer.parseInt(p2[1]));
            });
            for (String t : sortedTimes) {
                String[] p = t.split("-");
                timeKeys.add(new int[]{Integer.parseInt(p[0]), Integer.parseInt(p[1])});
            }
        }

        if (timeKeys.size() < 2) {
            throw new IllegalArgumentException("Нужно минимум 2 общих месяца данных.");
        }

        Map<String, Double> totalContrib = new HashMap<>();
        for (String v : drivers) totalContrib.put(v, 0.0);

        final double EPS = 1e-8;

        // 7. Инициализация GraalVM Polyglot Context
        try (Context context = Context.newBuilder("js").build()) {

            String jsFunctionCode = "(" + String.join(", ", drivers) + ") => (" + safeRight + ");";
            Value evalFunc = context.eval("js", jsFunctionCode);

            // 9. Помесячный цикл
            for (int i = 1; i < timeKeys.size(); i++) {
                int[] tPrev = timeKeys.get(i - 1);
                int[] tCurr = timeKeys.get(i);

                Map<String, Double> valsPrev = new HashMap<>();
                Map<String, Double> valsCurr = new HashMap<>();
                for (String v : drivers) {
                    valsPrev.put(v, safeData.get(v).get(tPrev[0]).get(tPrev[1]));
                    valsCurr.put(v, safeData.get(v).get(tCurr[0]).get(tCurr[1]));
                }

                // Готовим массивы аргументов для быстрого вызова JS функции по порядку аргументов
                Object[] argsPrev = new Object[drivers.size()];
                Object[] argsCurr = new Object[drivers.size()];
                for (int d = 0; d < drivers.size(); d++) {
                    argsPrev[d] = valsPrev.get(drivers.get(d));
                    argsCurr[d] = valsCurr.get(drivers.get(d));
                }

                // Вызов скомпилированной функции (работает мгновенно)
                double yPrev = evalFunc.execute(argsPrev).asDouble();
                double yCurr = evalFunc.execute(argsCurr).asDouble();
                double actualDelta = yCurr - yPrev;

                Map<String, Double> midpoints = new HashMap<>();
                Map<String, Double> deltas = new HashMap<>();
                for (String v : drivers) {
                    double p = valsPrev.get(v);
                    double c = valsCurr.get(v);
                    midpoints.put(v, (p + c) / 2.0);
                    deltas.put(v, c - p);
                }

                Map<String, Double> rawContrib = new HashMap<>();

                // Базовый массив аргументов для средних точек
                Object[] baseMidArgs = new Object[drivers.size()];
                for (int d = 0; d < drivers.size(); d++) {
                    baseMidArgs[d] = midpoints.get(drivers.get(d));
                }

                for (int d = 0; d < drivers.size(); d++) {
                    String v = drivers.get(d);
                    double midVal = midpoints.get(v);
                    double eps = Math.max(EPS, Math.abs(midVal) * 1e-7);

                    // Делаем копию базового массива средних точек, чтобы временно изменить один драйвер
                    Object[] argsPlus = baseMidArgs.clone();
                    argsPlus[d] = midVal + eps;
                    double fPlus = evalFunc.execute(argsPlus).asDouble();

                    Object[] argsMinus = baseMidArgs.clone();
                    argsMinus[d] = midVal - eps;
                    double fMinus = evalFunc.execute(argsMinus).asDouble();

                    double deriv = (fPlus - fMinus) / (2 * eps);
                    rawContrib.put(v, deriv * deltas.get(v));
                }

                double sumRaw = 0.0;
                for (double val : rawContrib.values()) sumRaw += val;
                double residual = actualDelta - sumRaw;

                double sumAbsRaw = 0.0;
                for (double val : rawContrib.values()) sumAbsRaw += Math.abs(val);
                for (String v : drivers) {
                    double raw = rawContrib.get(v);
                    double adjusted;
                    if (sumAbsRaw > 1e-12) {
                        double weight = Math.abs(raw) / sumAbsRaw;
                        adjusted = raw + residual * weight;
                    } else {
                        adjusted = raw + residual / drivers.size();
                    }
                    totalContrib.put(v, totalContrib.get(v) + adjusted);
                }
            }
        }

        // 10. Обратный маппинг алиасов -> бизнес-имена
        Map<String, Double> absResult = new LinkedHashMap<>();
        for (Map.Entry<String, Double> entry : totalContrib.entrySet()) {
            absResult.put(aliasToName.get(entry.getKey()), entry.getValue());
        }

        // 11. Возврат результатов
        if (!returnPercent) {
            return absResult;
        }

        double totalAbs = 0.0;
        for (double v : absResult.values()) totalAbs += Math.abs(v);
        Map<String, Double> percentResult = new LinkedHashMap<>();
        for (Map.Entry<String, Double> entry : absResult.entrySet()) {
            percentResult.put(entry.getKey(), totalAbs < 1e-12 ? 0.0 : (entry.getValue() / totalAbs) * 100.0);
        }

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("absolute", absResult);
        result.put("percent", percentResult);
        return result;
    }

    public static void main(String[] args) {
        Map<String, Map<Integer, Map<Integer, Double>>> testData = new LinkedHashMap<>();

        testData.put("Процент_Маржи", new LinkedHashMap<>() {{
            put(2026, new LinkedHashMap<>() {{ put(1, 20.0); put(2, 22.0); }});
        }});
        testData.put("Оборот_Подразделения", new LinkedHashMap<>() {{
            put(2026, new LinkedHashMap<>() {{ put(1, 5000000.0); put(2, 6000000.0); }});
        }});

        String formula = "Чистый_Доход = Процент_Маржи % Оборот_Подразделения";

        Object result = computeDrivers(formula, testData, false);
        System.out.println("Расчёт вкладов драйверов (Оптимизированный GraalVM Polyglot):");
        System.out.println(result);
    }
}

File: analytics/drivers.py
import sympy as sp
import re
from collections import defaultdict


def compute_drivers(formula, data, return_percent=False):
    left_str, right_str = formula.split('=')
    target_name = left_str.strip()

    name_to_alias = {}
    alias_to_name = {}
    for idx, name in enumerate(data.keys()):
        alias = f"VAR_{idx}"
        name_to_alias[name] = alias
        alias_to_name[alias] = name

    if target_name not in name_to_alias:
        name_to_alias[target_name] = "TARGET_VAR"
        alias_to_name["TARGET_VAR"] = target_name

    sorted_names = sorted(name_to_alias.keys(), key=len, reverse=True)
    pattern = re.compile("|".join(re.escape(n) for n in sorted_names))
    safe_right = pattern.sub(lambda m: name_to_alias[m.group(0)], right_str)

    safe_data = {name_to_alias[name]: ts for name, ts in data.items()}

    expr = sp.parse_expr(safe_right.strip())
    drivers = [str(s) for s in expr.free_symbols]
    if not drivers:
        raise ValueError("В формуле нет переменных.")

    derivatives = {var: sp.diff(expr, var) for var in drivers}
    ordered_symbols = [sp.Symbol(v) for v in drivers]

    eval_expr = sp.lambdify(ordered_symbols, expr, 'math')
    eval_diffs = {var: sp.lambdify(ordered_symbols, derivatives[var], 'math') for var in drivers}

    first_driver = drivers[0]
    time_keys = []
    for year in sorted(safe_data[first_driver].keys()):
        for month in sorted(safe_data[first_driver][year].keys()):
            time_keys.append((year, month))

    total_contrib = defaultdict(float)

    for i in range(1, len(time_keys)):
        t_prev = time_keys[i-1]
        t_curr = time_keys[i]

        vals_prev = {v: safe_data[v][t_prev[0]][t_prev[1]] for v in drivers}
        vals_curr = {v: safe_data[v][t_curr[0]][t_curr[1]] for v in drivers}

        args_prev = [vals_prev[v] for v in drivers]
        args_curr = [vals_curr[v] for v in drivers]

        y_prev = eval_expr(*args_prev)
        y_curr = eval_expr(*args_curr)
        actual_delta = y_curr - y_prev

        midpoints = {v: (vals_prev[v] + vals_curr[v]) / 2.0 for v in drivers}
        args_mid = [midpoints[v] for v in drivers]

        raw_contrib = {}
        for v in drivers:
            deriv = eval_diffs[v](*args_mid)
            raw_contrib[v] = deriv * (vals_curr[v] - vals_prev[v])

        sum_raw = sum(raw_contrib.values())
        residual = actual_delta - sum_raw

        sum_abs_raw = sum(abs(raw_contrib[v]) for v in drivers)
        if sum_abs_raw > 1e-12:
            for v in drivers:
                weight = abs(raw_contrib[v]) / sum_abs_raw
                total_contrib[v] += raw_contrib[v] + residual * weight
        else:
            for v in drivers:
                total_contrib[v] += raw_contrib[v] + residual / len(drivers)

    abs_result = {alias_to_name[alias]: contrib for alias, contrib in total_contrib.items()}

    if not return_percent:
        return abs_result

    total_abs = sum(abs(v) for v in abs_result.values())
    if total_abs < 1e-12:
        percent_result = {k: 0.0 for k in abs_result}
    else:
        percent_result = {k: (v / total_abs) * 100.0 for k, v in abs_result.items()}

    return {
        'absolute': abs_result,
        'percent': percent_result
    }


# ------------------------------------------------------------
# Пример
# ------------------------------------------------------------
# Вспомогательная функция для расчёта дохода по трём факторам
def calc_income(clients, check, conversion):
    return clients * check * conversion

if __name__ == "__main__":
    import random
    random.seed(42)

    # ------------------- ТЕСТ 1: РОСТ ДОХОДА -------------------
    print("=" * 60)
    print("ТЕСТ 1: РОСТ ДОХОДА (2024–2026)")
    print("=" * 60)

    years = [2024, 2025, 2026]
    months = list(range(1, 13))

    base_clients = 1000
    base_check = 500
    base_conversion = 0.80

    clients_growth = 0.05
    check_growth = 0.03
    conversion_growth = 0.02

    seasonal = [0.95, 0.90, 0.92, 0.98, 1.05, 1.10, 1.08, 1.02, 0.97, 0.96, 1.02, 1.12]

    test_data_up = {
        'Кол-во клиентов': {},
        'Средний чек': {},
        'Конверсия (%)': {}
    }

    for y in years:
        year_factor = 1 + (y - 2024) * 0.05
        for m in months:
            noise_clients = 1 + random.uniform(-0.02, 0.02)
            noise_check = 1 + random.uniform(-0.02, 0.02)
            noise_conversion = 1 + random.uniform(-0.02, 0.02)

            clients = base_clients * year_factor * seasonal[m-1] * noise_clients
            check = base_check * (1 + check_growth)**(y-2024) * (1 + 0.01 * (m-1)/12) * noise_check
            conversion = base_conversion * (1 + conversion_growth)**(y-2024) * (1 + 0.005 * (m-1)/12) * noise_conversion

            test_data_up['Кол-во клиентов'].setdefault(y, {})[m] = round(clients, 0)
            test_data_up['Средний чек'].setdefault(y, {})[m] = round(check, 2)
            test_data_up['Конверсия (%)'].setdefault(y, {})[m] = round(conversion, 4)

    formula = "Доход = Кол-во клиентов * Средний чек * Конверсия (%)"
    result_up = compute_drivers(formula, test_data_up, return_percent=True)

    first_year = min(test_data_up['Кол-во клиентов'].keys())
    first_month = min(test_data_up['Кол-во клиентов'][first_year].keys())
    last_year = max(test_data_up['Кол-во клиентов'].keys())
    last_month = max(test_data_up['Кол-во клиентов'][last_year].keys())

    income_first_up = calc_income(
        test_data_up['Кол-во клиентов'][first_year][first_month],
        test_data_up['Средний чек'][first_year][first_month],
        test_data_up['Конверсия (%)'][first_year][first_month]
    )
    income_last_up = calc_income(
        test_data_up['Кол-во клиентов'][last_year][last_month],
        test_data_up['Средний чек'][last_year][last_month],
        test_data_up['Конверсия (%)'][last_year][last_month]
    )

    print(f"Период: {first_month}.{first_year} – {last_month}.{last_year}")
    print(f"Доход в начале: {income_first_up:,.2f} руб.")
    print(f"Доход в конце:   {income_last_up:,.2f} руб.")
    print(f"Изменение дохода: {income_last_up - income_first_up:+,.2f} руб.\n")

    print("АБСОЛЮТНЫЕ ВКЛАДЫ (руб.):")
    for k, v in result_up['absolute'].items():
        print(f"  {k}: {v:+,.2f}")

    print("\nПРОЦЕНТНЫЕ ВКЛАДЫ (доля в изменении дохода):")
    for k, v in result_up['percent'].items():
        print(f"  {k}: {v:+.1f}%")

    sum_percent_up = sum(result_up['percent'].values())
    print(f"\nСумма процентов: {sum_percent_up:+.1f}%  (должно быть 100%)")

    sum_abs_up = sum(result_up['absolute'].values())
    delta_income_up = income_last_up - income_first_up
    print(f"\nБАЛАНС: сумма вкладов = {sum_abs_up:+.2f} руб., изменение дохода = {delta_income_up:+.2f} руб.")
    print("✅ Баланс сошёлся!" if abs(sum_abs_up - delta_income_up) < 1e-6 else "❌ Ошибка баланса!")

    # ------------------- ТЕСТ 2: ПАДЕНИЕ ДОХОДА -------------------
    print("\n" + "=" * 60)
    print("ТЕСТ 2: ПАДЕНИЕ ДОХОДА (2024–2026)")
    print("=" * 60)

    # Генерируем данные с падением
    clients_growth_fall = -0.05
    check_growth_fall = -0.03
    conversion_growth_fall = 0.02  # растёт, но не спасает

    test_data_down = {
        'Кол-во клиентов': {},
        'Средний чек': {},
        'Конверсия (%)': {}
    }

    for y in years:
        year_factor = 1 + (y - 2024) * clients_growth_fall
        for m in months:
            noise_clients = 1 + random.uniform(-0.02, 0.02)
            noise_check = 1 + random.uniform(-0.02, 0.02)
            noise_conversion = 1 + random.uniform(-0.02, 0.02)

            clients = base_clients * year_factor * seasonal[m-1] * noise_clients
            check = base_check * (1 + check_growth_fall)**(y-2024) * (1 + 0.01 * (m-1)/12) * noise_check
            conversion = base_conversion * (1 + conversion_growth_fall)**(y-2024) * (1 + 0.005 * (m-1)/12) * noise_conversion

            test_data_down['Кол-во клиентов'].setdefault(y, {})[m] = round(clients, 0)
            test_data_down['Средний чек'].setdefault(y, {})[m] = round(check, 2)
            test_data_down['Конверсия (%)'].setdefault(y, {})[m] = round(conversion, 4)

    result_down = compute_drivers(formula, test_data_down, return_percent=True)

    first_year = min(test_data_down['Кол-во клиентов'].keys())
    first_month = min(test_data_down['Кол-во клиентов'][first_year].keys())
    last_year = max(test_data_down['Кол-во клиентов'].keys())
    last_month = max(test_data_down['Кол-во клиентов'][last_year].keys())

    income_first_down = calc_income(
        test_data_down['Кол-во клиентов'][first_year][first_month],
        test_data_down['Средний чек'][first_year][first_month],
        test_data_down['Конверсия (%)'][first_year][first_month]
    )
    income_last_down = calc_income(
        test_data_down['Кол-во клиентов'][last_year][last_month],
        test_data_down['Средний чек'][last_year][last_month],
        test_data_down['Конверсия (%)'][last_year][last_month]
    )

    print(f"Период: {first_month}.{first_year} – {last_month}.{last_year}")
    print(f"Доход в начале: {income_first_down:,.2f} руб.")
    print(f"Доход в конце:   {income_last_down:,.2f} руб.")
    print(f"Изменение дохода: {income_last_down - income_first_down:+,.2f} руб.\n")

    print("АБСОЛЮТНЫЕ ВКЛАДЫ (руб.):")
    for k, v in result_down['absolute'].items():
        print(f"  {k}: {v:+,.2f}")

    print("\nПРОЦЕНТНЫЕ ВКЛАДЫ (доля в изменении дохода):")
    for k, v in result_down['percent'].items():
        print(f"  {k}: {v:+.1f}%")

    sum_percent_down = sum(result_down['percent'].values())
    print(f"\nСумма процентов: {sum_percent_down:+.1f}%  (должно быть -100%)")

    sum_abs_down = sum(result_down['absolute'].values())
    delta_income_down = income_last_down - income_first_down
    print(f"\nБАЛАНС: сумма вкладов = {sum_abs_down:+.2f} руб., изменение дохода = {delta_income_down:+.2f} руб.")
    print("✅ Баланс сошёлся!" if abs(sum_abs_down - delta_income_down) < 1e-6 else "❌ Ошибка баланса!")